tests/utils/managed_utils/mod.rs

```diff
@@ -0,0 +1,583 @@
+        let mut member_ids = Vec::new();
+        member_ids.push((0, group_creator_id));
+        let group = Group {
+            group_id: group_id.clone(),
+            members: member_ids,
+            ciphersuite: ciphersuite.clone(),
+            group_config: self.default_mgc.clone(),
+            public_tree,
+        };
+        groups.insert(group_id.clone(), group);
+        Ok(group_id)
+    }
+
+    /// Create a random group of size `group_size` and return the `GroupId`
+    pub fn create_random_group(
+        &self,
+        target_group_size: usize,
+        ciphersuite: &Ciphersuite,
+    ) -> Result<GroupId, SetupError> {
+        let group_id = self.create_group(ciphersuite)?;
+
+        let mut groups = self.groups.borrow_mut();
+        let group = groups.get_mut(&group_id).unwrap();
+
+        let new_members = self.random_new_members_for_group(group, target_group_size - 1)?;
+
+        let mut key_packages = Vec::new();
+        let clients = self.clients.borrow();
+        for member_id in &new_members {
+            let member = clients.get(member_id).unwrap().borrow();
+            key_packages.push(self.get_fresh_key_package(&member, ciphersuite)?);
+        }
+
+        while !key_packages.is_empty() {
+            // Pick a random adder.
+            let adder_id = group.random_group_member();
+            let adder = clients.get(&adder_id).unwrap().borrow();
+            let mut adder_group_states = adder.groups.borrow_mut();
+            let adder_group_state = adder_group_states.get_mut(&group_id).unwrap();
+            // Add between 1 and 5 new members.
+            let number_of_adds = ((OsRng.next_u32() as usize) % 5 % key_packages.len()) + 1;
+            let key_packages_to_add = key_packages.drain(0..number_of_adds);
+            let (mls_messages, welcome) = adder_group_state
+                .add_members(key_packages_to_add.as_slice())
+                .unwrap();
+            drop(adder_group_state);
+            drop(adder_group_states);
+            drop(adder);
+            self.distribute_to_members(&adder_id, group, mls_messages)?;
+            self.deliver_welcome(welcome, group)?;
+        }
+        Ok(group_id)
+    }
+
+    pub fn self_update(
+        &self,
+        action_type: ActionType,
+        group: &mut Group,
+        client_id: &Vec<u8>,
+    ) -> Result<(), SetupError> {
+        let clients = self.clients.borrow();
+        let client = clients
+            .get(client_id)
+            .ok_or(SetupError::UnknownClientId)?
+            .borrow();
+        let mut member_groups = client.groups.borrow_mut();
+        let member_group_state = member_groups
+            .get_mut(&group.group_id)
+            .ok_or(SetupError::ClientNotInGroup)?;
+        let (messages, welcome_option) = match action_type {
+            ActionType::Commit => {
+                // Let the function generate the key package bundle.
+                member_group_state.self_update(None)?
```

Comment:
So no way to test with an externally generated key package bundle?

Reply:
Good point, we should probably allow that.

Reply:
Done!